Add developer documentation for the autotuner (#284)
codeplaysoftware/sycl-blas: doc/Autotuner.md

# SYCL-BLAS Autotuner Developer Documentation

> This documentation aims to cover the inner workings of the auto tuner from a developer perspective.

Reviewer: ```suggestion
This documentation aims to cover the inner workings of the autotuner from a developer perspective.
```

The spelling `autotuner` seems more common in this PR.

> It should be useful if you need to work on or extend the autotuner in future, or just want to better understand how it works.

Reviewer: ```suggestion
It should be useful if you need to work on or extend the autotuner in the future, or just want to better understand how it works.
```

For information on building and using the autotuner please see the readme [here](../tools/auto_tuner/README.md).
# Table of Contents
- [**Overview**](#overview)
- [**C++ Binaries**](#c++-binaries)
- [**Parameter Configurations**](#parameter-configurations)
- [**Python Generation**](#python-generation)
    - [generated_combinations.def](#generated_combinations.def)
    - [Instantiation Source Files](#instantiation-source-files)
- [**CMake**](#cmake)
- [**Common Tasks**](#common-tasks)
    - [Adding a new set of configurations](#adding-a-new-set-of-configurations)

# Overview
The Autotuner is comprised of several parts:
- The `C++ binaries` that are run by the user to do the tuning. 

> Comprised of 5 different binaries which benchmark different combinations of transposition of the inputs and print the results.

Reviewer: I see there's a good reason for there being 5 binaries, but I think we still need to be careful since documentation like this can quickly go out of date. Maybe try to unify the way it's referred to so that it's easier to grep for:
```suggestion
Comprised of five different binaries which benchmark different combinations of transposition of the inputs and print the results.
```
However this is a minor concern for this specific PR.

The fifth binary (`tune_all`) runs all four combinations one after the other.

The bulk of the work is done inside `gemm_tuner.hpp` in the `run_tune_gemm()` function. 
This function is called by each `tune_` binary with different combinations of template parameters for transposition.
Before testing the generated combinations the application first benchmarks the provided `system BLAS`, along with the current `SYCL-BLAS`. 
The generated combinations are then run with the following macro:
```c++
#define BENCH_PARAMS(MEM, ALG, BATCH, VEC, ...)                             \
  do {                                                                      \
    auto result =                                                           \
        tune<__VA_ARGS__, GemmConfig<TransA, TransB, MEM, ALG, BATCH, VEC>, \
             DataType>(rep, args);                                          \
    results.push_back(result);                                              \
  } while (0);

#include "generated_combinations.def"

#undef BENCH_PARAMS
```

> The file `generated_combinations.def` contains calls to `BENCH_PARAMS` for each combination to be tested, and the stored results are sorted from worst to best before being printed. For more information on this file and how it is generated, see [this section](#generated_combinations.def).

Reviewer: ```suggestion
The file `generated_combinations.def` contains calls to `BENCH_PARAMS` for each combination to be tested,
and the stored results are sorted from worst to best before being printed.
For more information on this file and how it is generated, see [this section](#generated_combinations.def).
```

There are several wrapper functions which wrap the function `run_tune()`: `tune()` and `tune_syclblas()`. 
The `run_tune()` function takes a function object, runs it and measures the execution time. 
The `tune()` wrapper function is used in the above macro, and `tune_syclblas()` is used for testing the current SYCL-BLAS. 

Calling `run_tune()` happens directly in `run_tune_gemm()` for testing the system BLAS.

# Parameter Configurations
Parameter configurations are provided as JSON files. 
The autotuner includes configs for various target backends along with a set of default configs.

Detailed information about the config parameters can be found in the autotuner's [readme](../tools/auto_tuner/README.md#configuration).
# Python Generation
The python script `tools/auto_tuner/gen/generate_combinations.py` produces two types of output:

> - A single file, `generated_combinations.def`, which contains generated calls to the `BENCH_PARAMS` macro. As seen earlier this file is included inside the `run_tune_gemm()` function after the `BENCH_PARAMS` macro is defined. This handles the execution of each test configuration.

Reviewer: ```suggestion
- A single file, `generated_combinations.def`, which contains generated calls to the `BENCH_PARAMS` macro.
  As seen earlier this file is included inside the `run_tune_gemm()` function after the `BENCH_PARAMS` macro is defined.
  This handles the execution of each test configuration.
```

> - A number of `C++` source files which instantiate calls to the `tune<>` method, with one file for each configuration. This keeps the kernels for each configuration in separate translation units.

Reviewer: ```suggestion
- A number of `C++` source files which instantiate calls to the `tune<>` method, with one file for each configuration.
  This keeps the kernels for each configuration in separate translation units.
```

  VERBATIM
)
add_custom_target(tuner_generate_def_file DEPENDS ${tuner_def_file})
```
First the `generated_combinations.def` file is generated (see [the previous section](#generated_combinations.def) for more information). This file is then added to a custom target which is later added as a dependency for each C++ binary later on.
```CMake
set(_gen_src_dir ${CMAKE_CURRENT_BINARY_DIR}/gen)
# older version of python cannot generate the dir when it does not exist
file(MAKE_DIRECTORY ${_gen_src_dir})
execute_process(
  COMMAND ${PYTHON_EXECUTABLE} ${_generator_script}
    ${GEN_CONFIG} --source_dir=${_gen_src_dir} --list_files
  OUTPUT_VARIABLE _gen_src_files
  RESULT_VARIABLE _gen_src_files_exitcode
  OUTPUT_STRIP_TRAILING_WHITESPACE
)
```
Then the script is run once more with the `--list_files` option passed. This will output a list of the source files that will be generated in the next step.
```CMake
add_custom_command(OUTPUT ${tuner_kernel_srcs}
  COMMAND ${PYTHON_EXECUTABLE} ${_generator_script}
    ${GEN_CONFIG} --source_dir=${_gen_src_dir}
  MAIN_DEPENDENCY ${GEN_CONFIG}
  DEPENDS ${_generator_script}
)
add_custom_target(tuner_generate_kernels DEPENDS ${tuner_kernel_srcs})
```

> The third execution of the script generates the source file instantiations (covered in [this section](#instantiation-source-files)) and adds the list of sources to another custom target. The source files are added as a library, `tuner_kernel_lib` and the `tuner_generate_kernels` target is added as a dependency of this library.

Reviewer: ```suggestion
The third execution of the script generates the source file instantiations (covered in [this section](#instantiation-source-files))
and adds the list of sources to another custom target.
The source files are added as a library, `tuner_kernel_lib` and the `tuner_generate_kernels` target is added as a dependency of this library.
```
